swift/extractor/SwiftDispatcher.h

```diff
@@ -0,0 +1,178 @@
+  case swift::TypeKind::CLASS: \
+    return #CLASS;
+#include "swift/AST/TypeNodes.def"
+    default:
+      return "Unknown";
+  }
+}
+
+template <>
+std::string inline getKindName<swift::TypeRepr, swift::TypeReprKind>(swift::TypeReprKind kind) {
+  switch (kind) {
+#define TYPEREPR(CLASS, PARENT)    \
+  case swift::TypeReprKind::CLASS: \
+    return #CLASS;
+#include "swift/AST/TypeReprNodes.def"
+    default:
+      return "Unknown";
+  }
+}
+
+}  // namespace detail
+
+// The main reponsibilities of the SwiftDispatcher are as follows:
+// * redirect specific AST node emission to a corresponding visitor (statements, expressions, etc.)
+// * storing TRAP labels for emitted AST nodes (in the TrapLabelStore) to avoid re-emission
+// Since SwiftDispatcher sees all the AST nodes, it also attaches a location to every 'locatable'
+// node (AST nodes that are not types: declarations, statements, expressions, etc.).
+class SwiftDispatcher {
+ public:
+  SwiftDispatcher(const swift::SourceManager& sourceManager, TrapArena& arena, TrapOutput& trap)
+      : sourceManager{sourceManager}, arena{arena}, trap{trap} {}
+
+  template <typename T>
+  void extract(T* entity) {
+    fetchLabel(entity);
+  }
+
+ private:
+  // This method gives a TRAP label for already emitted AST node.
+  // If the AST node was not emitted yet, then the emission is dispatched to a corresponding
+  // visitor (see `visit(T *)` methods below).
+  template <typename E>
+  TrapLabel<ToTag<E>> fetchLabel(E* e) {
+    // this is required so we avoid any recursive loop: a `fetchLabel` during the visit of `e` might
+    // end up calling `fetchLabel` on `e` itself, so we want the visit of `e` to call `fetchLabel`
+    // only after having called `assignNewLabel` on `e`
+    assert(!waitingForNewLabel && "fetchLabel called before assignNewLabel");
+    if (auto l = store.get(e)) {
+      return *l;
+    }
+    waitingForNewLabel = getCanonicalPtr(e);
+    visit(e);
+    if (auto l = store.get(e)) {
+      if constexpr (!std::is_base_of_v<swift::TypeBase, E>) {
+        attachLocation(e, *l);
+      }
+      return *l;
+    }
+    assert(!"assignNewLabel not called during visit");
+    return {};
+  }
+
+  // Due to the lazy emission approach, we must assign a label to a corresponding AST node before
+  // it actually gets emitted to handle recursive cases such as recursive calls, or recursive type
+  // declarations
+  template <typename E>
+  TrapLabel<ToTag<E>> assignNewLabel(E* e) {
+    assert(waitingForNewLabel == getCanonicalPtr(e) && "assignNewLabel called on wrong entity");
+    auto label = getLabel<ToTag<E>>();
+    trap.assignStar(label);
+    store.insert(e, label);
+    waitingForNewLabel = nullptr;
+    return label;
+  }
+
+  template <typename Tag>
+  TrapLabel<Tag> getLabel() {
+    return arena.allocateLabel<Tag>();
+  }
+
+  // This is a helper method to emit TRAP entries for AST nodes that we don't fully support yet.
+  template <typename Parent, typename Child>
```

Comment:
(Probably not in this PR - let's file an issue?)

There are a few places where we explicitly pass a `Parent` or have to check that it's a base. If I understand the logic correctly, this is always one of a fixed set of base classes, and there is always a unique match. Would it make sense to instead add a metafunction that computes the parent automatically based on the rule that whichever of the fixed set is *a* parent is *the* parent?

Reply:
What would be the goal to add such a metafunction? To eliminate the explicit `Parent` type argument at the call sites?
If that's the only motivation, then I don't think it is worth the effort - we have 6 hard-coded calls and they are very unlikely to change in the future.

Reply:
probably yes, although this will look like something like the following much simpler code after we discuss and merge changes proposed in https://github.com/github/semmle-code/pull/42517:

```cpp
  template <typename E>
  void emitUnknown(E* entity) {
    auto label = assignNewLabel(entity);
    emit(trap::TagToBindingTrap<typename decltype(label)::Tag>{label});
    emit(trap::ElementIsUnknown{label});
  }
```